elasticsearch.go

```diff
@@ -189,12 +218,107 @@ func NewClient(cfg Config) (*Client, error) {
 		go client.DiscoverNodes()
 	}
 
-	return client, nil
+	if cfg.UseHeaderCheckOnly {
+		useHeaderCheckOnly = true
+	}
+
+
+	return client, err
+}
+
+func genuineCheckHeader(header http.Header) error {
+	if header.Get("X-Elastic-Product") != "Elasticsearch" {
+		return fmt.Errorf(unknownProduct)
+	}
+	return nil
+}
+
+func genuineCheckInfo(info info) error {
+	major, minor, _, err := ParseElasticsearchVersion(info.Version.Number)
+	if err != nil {
+		return err
+	}
+
+	if major < 6 {
+		return fmt.Errorf(unknownProduct)
+	}
+	if major < 7 {
+		if info.Tagline != tagline {
+			return fmt.Errorf(unknownProduct)
+		}
+	}
+	if major >= 7 {
+		if minor < 14 {
+			if info.Tagline != tagline ||
+				info.Version.BuildFlavor != "default" {
+				return fmt.Errorf(unknownProduct)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ParseElasticsearchVersion returns an int64 representation of Elasticsearch version.
+//
+func ParseElasticsearchVersion(version string) (int64, int64, int64, error) {
+	matches := reVersion.FindStringSubmatch(version)
+
+	if len(matches) < 4 {
+		return 0, 0, 0, fmt.Errorf("")
+	}
+	major, _ := strconv.ParseInt(matches[1], 10, 0)
+	minor, _ := strconv.ParseInt(matches[2], 10, 0)
+	patch, _ := strconv.ParseInt(matches[3], 10, 0)
+
+	return major, minor, patch, nil
 }
 
 // Perform delegates to Transport to execute a request and return a response.
 //
 func (c *Client) Perform(req *http.Request) (*http.Response, error) {
+	if !productChecked {
+		var info info
+
+		res, err := c.Info()
+		if err != nil {
+			return nil, err
+		}
+
+		contentType := res.Header.Get("Content-Type")
+		if res.Body != nil {
+			defer res.Body.Close()
+
+			if strings.Contains(contentType, "json") {
+				decoder := json.NewDecoder(res.Body)
+				err = decoder.Decode(&info)
+				if err != nil {
+					return nil, fmt.Errorf("error decoding Elasticsearch informations: %s", err)
+				}
+			}
+
+			switch res.StatusCode {
+			case http.StatusForbidden:
+			case http.StatusUnauthorized:
+				break
+			default:
+				err = genuineCheckHeader(res.Header)
+
+				if err != nil {
+					if !useHeaderCheckOnly && info.Version.Number != "" {
```

Comment:
Do we need to special case the empty string for `info.Version.Number` if it'll be handled within `genuineCheckInfo()`?

Reply:
In Go if the structure is not initialized it's field have default type values, in this case `string` is just an empty string.